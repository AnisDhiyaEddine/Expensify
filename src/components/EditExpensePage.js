import React from "react";

const EditExpensePage = (props) => {
console.log(props.match)
    return (<div>This is from the edit expense page</div>)
}

export default EditExpensePage;
