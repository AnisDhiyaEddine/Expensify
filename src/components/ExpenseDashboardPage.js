import React from "react";


const ExpenseDashboardPage = () => (
  <div>This is from the app dashboard page</div>
);

export default ExpenseDashboardPage;
