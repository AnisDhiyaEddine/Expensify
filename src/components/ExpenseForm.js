import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.expense : "",
      amount: props.expense ? props.expense.amount : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocus: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      e.target.value = amount;

      this.setState(() => ({
        amount: parseFloat(amount) || 0,
      }));
    }
  };

  onDateChange = (createdAt) => {
    this.setState(() => ({
      createdAt,
    }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocus: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (this.state.description && this.state.amount) {
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
      });
      this.setState(() => ({
        error: "",
      }));
    } else {
      this.setState(() => ({
        error: "Please provide the description and the amount",
      }));
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocus}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (Optional"
            defaultValue={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
