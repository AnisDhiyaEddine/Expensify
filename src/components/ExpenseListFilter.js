import React from "react";
import { connect } from "react-redux";
import { v4 } from "uuid";
import { setTextFilter, sortByAmount, sortByDate } from "../actions/filters";
import moment from "moment";
import { DateRangePicker } from "react-dates";
import "react-dates/initialize";
import { setStartDate, setEndDate } from "../actions/filters";

export class ExpenseListFilter extends React.Component {
  state = {
    startDate: moment().startOf("month"),
    endDate: moment().endOf("month"),
    focusedInput: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);

    this.setState(() => ({
      startDate,
      endDate,
    }));
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({
      focusedInput,
    }));
  };

  onTextChange = (event) => {
    this.props.setTextFilter(event.target.value);
  };

  onSortByChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <input
          type="text"
          defaultValue={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select
          defaultValue={this.props.filters.sortBy}
          onChange={this.onSortByChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          className="DateRangePicker"
          startDate={this.state.startDate}
          endDate={this.state.endDate}
          endDateId={v4()}
          startDateId={v4()}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
