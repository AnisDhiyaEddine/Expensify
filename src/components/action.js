import React from "react";

const Action = (props) => {
  return (
    <div>
      <button
        className="big_button"
        disabled={props.disable}
        onClick={props.pickOption}
      >
        What should i do
      </button>
    </div>
  );
};

export default Action;
