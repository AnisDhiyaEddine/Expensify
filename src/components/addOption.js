import React from "react";

export default class AddOption extends React.Component {
  state = {
    error: undefined,
  };

  addOption = (event) => {
    event.preventDefault();
    this.setState(() => ({ error: "" }));

    let option = event.target.elements.addOption.value;
    event.target.elements.addOption.value = "";
    let error = this.props.addOption(option);
    if (error) {
      console.log(error);
      this.setState(() => ({ error }));
    }
  };
  render() {
    return (
      <div>
        <p className="add-option-error">{this.state.error}</p>
        <form className="add-option" onSubmit={this.addOption}>
          <input className="add-option__input" name="addOption" />
          <button className="button">Add option</button>
        </form>
      </div>
    );
  }
}
