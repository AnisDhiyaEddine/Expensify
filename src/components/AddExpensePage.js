import React from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";
import { connect } from "react-redux";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>his is from the add expense page</h1>
        <ExpenseForm className="ExpenseForm" onSubmit={this.onSubmit} />
      </div>
    );
  }
}

let mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
