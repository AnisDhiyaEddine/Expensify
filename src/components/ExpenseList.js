import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpences from "../selectors/expenses";
import ExpenseListFilter from "./ExpenseListFilter";

const ExpenseList = (props) => (
  <div>
    {<ExpenseListFilter />}
    <h1> Expenses list</h1>
    {props.expenses.map((expense) => (
      <ExpenseListItem key={expense.id} {...expense} />
    ))}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpences(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
