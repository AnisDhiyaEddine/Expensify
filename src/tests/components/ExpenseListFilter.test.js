import React from "react";
import { shallow } from "enzyme";
import { ExpenseListFilter } from "../../components/ExpenseListFilter";
import { filters, altFilters } from "../fixtures/filters";
import moment from "moment";
let wrapper, setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByAmount = jest.fn();
  sortByDate = jest.fn();

  setStartDate = jest.fn();
  setEndDate = jest.fn();
  wrapper = shallow(
    <ExpenseListFilter
      setTextFilter={setTextFilter}
      sortByAmount={sortByAmount}
      sortByDate={sortByDate}
      setStartDate={setStartDate}
      setEndDate={setEndDate}
      filters={filters}
    />
  );
});

test("render ExpenseListFilter with default filters", () => {
  expect(wrapper).toMatchSnapshot();
});

test("render ExpenseListFilter with custom filters", () => {
  wrapper.setProps({ filters: altFilters });
  expect(wrapper).toMatchSnapshot();
});

//handle on text change

test("test the text change filter", () => {
  wrapper
    .find("input")
    .at(0)
    .simulate("change", {
      target: {
        value: "New text filter",
      },
    });
  expect(setTextFilter).toBeCalledWith("New text filter");
});

//sort by date
test("sortByDate", () => {
  wrapper.setProps({ filters: altFilters });

  wrapper.find("select").simulate("change", {
    target: {
      value: "date",
    },
  });
  expect(sortByDate).toHaveBeenCalled();
});

//sort by amount

test("sortByDate", () => {
  wrapper.find("select").simulate("change", {
    target: {
      value: "amount",
    },
  });
  expect(sortByAmount).toHaveBeenCalled();
});

//handle date changes
test("handle date changes", () => {
  let startDate = moment(0).add(2, "days");
  let endDate = moment(0).add(5, "days");
  wrapper.find(".DateRangePicker").prop("onDatesChange")({
    startDate,
    endDate,
  });

  expect(setStartDate).toHaveBeenCalledWith(moment(0).add(2, "days"));
  expect(setEndDate).toHaveBeenCalledWith(moment(0).add(5, "days"));
});
//handle date focus changes

test("handle date changes", () => {
  wrapper.find(".DateRangePicker").prop("onFocusChange")("second");

  expect(wrapper.state("focusedInput")).toBe("second");
});
