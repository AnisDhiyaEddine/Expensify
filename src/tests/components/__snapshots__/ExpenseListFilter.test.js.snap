// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseListFilter with custom filters 1`] = `
<div>
  <input
    defaultValue="darling"
    onChange={[Function]}
    type="text"
  />
  <select
    defaultValue="amount"
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    className="DateRangePicker"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={"1970-01-31T23:59:59.999Z"}
    endDateId="7d3fc2b3-0320-49fa-80e9-d77bac47692c"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={false}
    showDefaultInputIcon={false}
    small={false}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="6e41f0d0-ff9a-46db-a6d9-6fb640d0186a"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;

exports[`render ExpenseListFilter with default filters 1`] = `
<div>
  <input
    defaultValue=""
    onChange={[Function]}
    type="text"
  />
  <select
    defaultValue="date"
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    className="DateRangePicker"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={"1970-01-31T23:59:59.999Z"}
    endDateId="fc1a7080-c6f2-46bf-8a7e-e95ba11808dc"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={false}
    showDefaultInputIcon={false}
    small={false}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="4c916c4c-f5ae-44e1-be75-2f82d8a40896"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
